### DISCUSSION TITLE
adding typeset/declare/local to Zsh/Sh parser

### DISCUSSION THREAD

**Discussion by psprint** [universal-ctags/ctags#3665] [Q&A]
I'm looking for a guidance for adding local variables support to sh/zsh parsers. They're typically defined with a `typeset`/`declare`/`local` keyword. What is the way of adding this to universal-ctags?

The name of the parser:
Sh/Zsh

The tags output you expect:

A output that has shell variables tagged.

The version of ctags:

```
$ ctags --version
Universal Ctags 6.0.0(p6.0.20230219.0), Copyright (C) 2015-2022 Universal Ctags Team
Universal Ctags is derived from Exuberant Ctags.
Exuberant Ctags 5.8, Copyright (C) 1996-2009 Darren Hiebert
  Compiled: Feb 19 2023, 09:38:36
  URL: https://ctags.io/
  Output version: 0.0
  Optional compiled features: +wildcards, +regex, +iconv, +option-directory, +xpath, +packcc, +optscript, +pcre2
```

How do you get ctags binary:

I'm cloning and autotools building it.

**Comment by masatake** [comment]
About supporting local variables, a suggestion I can give is only "read sh.c".

To support local variables, the parser must support scopes of functions.

```
f()
{
   local x
   ...
}

g()
{
   ...
}
```

The current parser implementation doesn't know the end of f(). Extracting `x` is not hard, but it is not easy to tell in which function x is defined/used. We call such information "scope".
Knowing the end of a function is mostly the same as knowing its scope.

python.c tracks the scope cleanly (token based).
ruby.c tracks the scope in a dirty way (line oriented).


Compared with supporting scope, adding and supporting may be an easier task.
d30bac01d1 illustrates the way to add a keyword.

I struggled to add support for variables: https://github.com/universal-ctags/ctags/pull/3544 .
However, I didn't finish the work because I was not sure it was the way to go.

Good luck!

**Comment by masatake** [comment]
I think this is not an "issue". So if you are o.k., I would like to convert this to a discussion.
